src/dotnet/CarbonAware.DataSources.WattTime/WattTimeDataSource.cs

```diff
@@ -0,0 +1,74 @@
+﻿using CarbonAware.Interfaces;
+using CarbonAware.Model;
+using CarbonAware.Tools.WattTimeClient;
+using Microsoft.Extensions.Logging;
+using System.Diagnostics;
+using System.Globalization;
+
+namespace CarbonAware.DataSources.WattTime;
+
+/// <summary>
+/// Reprsents a wattime data source.
+/// </summary>
+public class WattTimeDataSource : ICarbonIntensityDataSource
+{
+    private ILogger<WattTimeDataSource> Logger { get; }
+
+    private IWattTimeClient WattTimeClient { get; }
+
+    private ActivitySource ActivitySource { get; }
+
+    private IRegionConverter RegionConverter { get; }
+
+    /// <summary>
+    /// Creates a new instance of the <see cref="WattTimeDataSource"/> class.
+    /// </summary>
+    /// <param name="logger">The logger for the datasource</param>
+    /// <param name="client">The WattTime Client</param>
+    /// <param name="activitySource">The activity source for telemetry.</param>
+    /// <param name="regionConverter">The region converter to be used to convert Azure regions to BA's.</param>
+    public WattTimeDataSource(ILogger<WattTimeDataSource> logger, IWattTimeClient client, ActivitySource activitySource, IRegionConverter regionConverter)
+    {
+        this.Logger = logger;
+        this.WattTimeClient = client;
+        this.ActivitySource = activitySource;
+        this.RegionConverter = regionConverter;
+    }
+
+    /// <inheritdoc />
+    public async Task<IEnumerable<EmissionsData>> GetCarbonIntensityAsync(string region, DateTimeOffset startPeriod, DateTimeOffset endPeriod)
+    {
+        this.Logger.LogInformation("Getting carbon intensity for region {region} for period {startPeriod} to {endPeriod}.", region, startPeriod, endPeriod);
+
+        using (var activity = ActivitySource.StartActivity())
+        {
+            var balancingAuthority = await this.RegionConverter.ConvertAzureRegionAsync(region);
+
+            if (balancingAuthority == null)
+            {
+                Logger.LogError("Unable to find a balancing authority for region {region}", region);
```

Comment:
nit: declare the string once with string.format and use it for both the log message and the exception.

Reply:
Actually, with structured logging, they're different.  String.Format does a string allocation and formats the string by replacing the value.  Structure logging is different.  You'll note that this is NOT a string formatting operation (there's no $ or string.format).  Instead, under the covers, we're creating a map that has "region" with a value of region inside of it.  It will, eventually, do a string operation to replace {region} with region.  If we use a string format operation here, we lose structured logging.